File: GenericSort.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <iomanip>
using namespace std;

//define class type for person
class Person {
    public: 
        string name;
        int age;
};

//override function for Person class to compare ages
bool ageCompare(const Person& a, const Person& b) {
    if (a.age != b.age) {
        return a.age > b.age;
    }
    return a.name < b.name;
}

//override function for Person class to compare names
bool nameCompare(const Person& a, const Person& b) {
    if (a.name != b.name) {
        return a.name < b.name;
    }
    return a.age < b.age;
}

//override function for printing an element from the Person class
ostream &operator<<(ostream &output, const Person &p) {
    output << p.name << ", " << p.age << " ";
    return output;
}

int main() {
    vector <float> arr = {645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25};
    int length = arr.size();
    // cout << length << endl;
    //use generic sort programming here to sort elements in the array
    sort(arr.begin(), arr.end());
    cout << "Comparing Float Numbers Here" << endl;
    for (int i = 0; i < length; i++) {
        cout << fixed << setprecision(2) << arr[i] << ", ";
    }
    cout << endl;
    cout << endl;

    vector <int> arr2 = {300, 10, 60, 9, 15, 16, 19, 30, 60, 50, 90, 100, 70, 80};
    int length2 = arr.size();
    // cout << length << endl;
    //use generic sort programming here to sort elements in the array
    sort(arr2.begin(), arr2.end());
    cout << "Comparing Int Numbers Here" << endl;
    for (int i = 0; i < length; i++) {
        cout << fixed << setprecision(2) << arr2[i] << ", ";
    }

    cout << endl;
    cout << endl;
    cout << "Comparing Person's Names Here" << endl;
    vector <Person> arr3 = {{"Hal", 20}, {"Susann", 31}, {"Dwight", 19}, {"Kassandra", 21}, {"Lawrence", 25}, {"Cindy", 22}, {"Cory", 27}, {"Mac", 19}, {"Romana", 27}, {"Doretha", 32}, {"Danna", 20}, {"Zara", 23}, {"Rosalyn", 26}, {"Risa", 24}, {"Benny", 28}, {"Juan", 33}, {"Natalie", 25}};
    int length3 = arr3.size();
    // cout << "length 3: "  << length3 << endl;
    //takes third parameter nameCompare to override the function and allows us to sort the person class by name
    sort(arr3.begin(), arr3.end(), nameCompare);
     for (int i = 0; i < length3; i++) {
        cout << arr3[i] << "; ";
    }

    cout << endl;
    cout << endl;
    cout << "Comparing Person's Ages Here" << endl;
    vector <Person> arr4 = {{"Hal", 20}, {"Susann", 31}, {"Dwight", 19}, {"Kassandra", 21}, {"Lawrence", 25}, {"Cindy", 22}, {"Cory", 27}, {"Mac", 19}, {"Romana", 27}, {"Doretha", 32}, {"Danna", 20}, {"Zara", 23}, {"Rosalyn", 26}, {"Risa", 24}, {"Benny", 28}, {"Juan", 33}, {"Natalie", 25}};
    //takes third parameter ageCompare to override the function and allows us to sort the person class by age
    sort(arr4.begin(), arr4.end(), ageCompare);
    int length4 = arr4.size();
    // cout << "length 4: "  << length4 << endl;
    for (int i = 0; i < length4; i++) {
        cout << arr4[i] << "; ";
    }
}

File: C#-files/GenericSort.cs
using System;
using System.Linq;

class Program {
  public static void Main (string[] args) {
    Console.WriteLine ("Hello Assignment 2\n");
    
    Console.WriteLine("persons BEFORE sorting");
    Person [] persons = {new Person("Hal", 20), new Person("Susann", 31), new Person("Dwight", 19), new Person("Kassandra", 21), new Person("Lawrence", 25), new Person("Cindy", 22), new Person("Cory", 27), new Person("Mac", 19), new Person("Romana", 27), new Person("Doretha", 32), new Person("Danna", 20), new Person("Zara", 23), new Person("Rosalyn", 26), new Person("Risa", 24), new Person("Benny", 28), new Person("Juan", 33), new Person("Natalie", 25)};
    
    foreach (Person p in persons){
      Console.WriteLine(p.name + ", " + p.age);
    }

    Console.WriteLine("\npersons AFTER sorting");
    persons = persons.OrderBy(p => p).ToArray();
    //Array.Sort(persons);
    foreach (Person p in persons){
      Console.WriteLine(p.name + ", " + p.age);
    }

    
    Console.WriteLine ("\n\nnumbers BEFORE sorting");
    double[] numbers = {645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25};
    foreach (double i in numbers){
      Console.Write(i + " ");
    }

    //numbers = numbers.OrderByDescending(x =>x).ToArray();
    Array.Sort(numbers);
    Console.WriteLine ("\nnumbers AFTER sorting");
    foreach (double i in numbers){
      Console.Write(i + " ");
    }
  }
}

public class Person : IComparable<Person> {
  public int age;
  public String name;

  public Person(String n, int a){
    age = a;
    name = n;
  }

   public int CompareTo(Person that)
    {
        if (this.age >  that.age) return -1;
        if (this.age == that.age) return this.name.CompareTo(that.name);;
        return 1;
    }
  
}